designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,499 @@
+1. It encourages plugin authors to use the `configs` key on their plugin in order to allow this usage.
+1. It allows ESLint to modify configs before they are used (see below).
+
+#### Reassignable Plugin Configs
+
+When using named configs, this gives ESLint the opportunity to inspect and change the config before applying it. This means we have the opportunity to improve the plugin config experience.
+
+Right now, bundling a config in a plugin is a bit of a messy process for two reasons:
+
+1. You need to include the plugin in the config directly, which means that you can't just have an single object literal defined as the plugin. You need to define the plugin first and then add the configs with the plugin reference second.
+1. You don't actually know what namespace a user will assign the plugin to, so while you might assume someone will import `eslint-plugin-import` as `import` and you then refer to rules from the plugin as `import/rule`, there is nothing preventing the user from assigning the plugin to a different name. In that case, the user won't know how to override rule options or assign different languages from the plugin.
+
+Here's an actual example from [`@eslint/json`](https://github.com/eslint/json):
+
+```js
+const plugin = {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+    // can't include the plugin here because it must reference `plugin`
+	configs: {},
+};
+
+Object.assign(plugin.configs, {
+	recommended: {
+
+        // now we can reference plugin
+		plugins: { json: plugin },
+		rules: {
+			"json/no-duplicate-keys": "error",
+			"json/no-empty-keys": "error",
+		},
+	},
+});
+
+export default plugin;
+```
+
+Here, we are hardcoding the namespace `json` even though that might not be the namespace that the user assigns to this plugin. This is something we can now address with the use of `extends` because we have the ability to alter the config before inserting it.
+
+Instead of using a hardcoded plugin namespace, plugins can instead use `#` to indicate that they'd like to have the plugin itself included and referenced using the namespace the user assigned. For example, we could rewrite the JSON plugin like this:
+
+```js
+export default {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+	configs: {
+        recommended: {
+            plugins: { "#": null },
+            rules: {
+                "#/no-duplicate-keys": "error",
```

Comment:
Wouldn't using this rewrite mechanism then disallow people to use the rules directly from plugin config objects?  Today, it's not uncommon to just spread a plugin's config's rules into a config.  In this model, if they did that, the rules wouldn't have the token replaced, right? If that's the case, it seems like taking that usage pattern away would be a big side effect.

Reply:
The intent isn't to take that usage pattern away, but rather to have another option. 

Also, people are accessing `rules` directly right now because there isn't another way to easily extend configs. I'm not sure how necessary that pattern is in this proposal.